Add extra support for meta fields

becoded opened the pull request:
Add 

- Created
- LastModified
- Version

di-wu commented on server.go:
```diff
@@ -60,7 +60,23 @@ func (s Server) getSchema(id string) schema.Schema {
 	w.Header().Set("Content-Type", "application/scim+json")
-	path := strings.TrimPrefix(r.URL.Path, "/v2")
+
+	path := r.URL.Path
+	basePath := strings.TrimSuffix(s.getBasePath(r), "/")
+
+	if !strings.HasPrefix(path, basePath) {
+		errorHandler(w, r, scimError{
+			detail: "Specified endpoint does not exist.",
+			status: http.StatusNotFound,
+		})
+
+		return
+	}
+
+	path = strings.TrimPrefix(path, basePath)
```
Is there a reason to do this within the SCIM Server and not before it gets passed to the `HandlerFunc`? I think we need to keep the server and other infrastructural components as separate as possible. I.e. if everything would be located after the endpoint `/api/users/scim/` then this should be stripped before being passed to the SCIM handler.

Any thoughts about this?

becoded replied:
The reason we added it was to be able to set the location of the `meta` correctly.

If you expose the endpoint as `/Users` then you get the following response
```
{
  "id": "2",
  "userName": "Jane Doe",
  "meta": {
    "resourceType": "User",
    "created": "2018-04-17T09:03:39-05:00",
    "lastModified": "2018-04-17T09:03:39-05:00",
    "location": "Users/2"
  },
  ...
```

If you expose the endpoint as `/api/users/scim/Users` then you get the same response, however, your location needs to be different like this:
```
{
  "id": "2",
  "userName": "Jane Doe",
  "meta": {
    "resourceType": "User",
    "created": "2018-04-17T09:03:39-05:00",
    "lastModified": "2018-04-17T09:03:39-05:00",
    "location": "api/users/scim/Users/2"
  },
  ...
```
So you need a way to prefix your location path.

Do you have a proposal to be able to have a correct location in a different way?

di-wu replied:
I think I interpreted the specification somewhat differently. Let me try to explain where I am coming from!

In the [RFC7643.3.1](https://tools.ietf.org/html/rfc7643#section-3.1) states that the `location` value must be the same as the "Content-Location" HTTP response header which is defined in [RFC7231.3.1.4.2](https://tools.ietf.org/html/rfc7231#section-3.1.4.2).

```
Content-Location = absolute-URI / partial-URI
```

In our case the value is a `partial-URI`  since we do not return the scheme and authority.

```
partial-URI   = relative-part [ "?" query ]
relative-part = "//" authority path-abempty 
    / path-absolute 
    / path-noscheme 
    / path-empty

path-abempty  = *( "/" segment )
path-absolute = "/" [ segment-nz *( "/" segment ) ]
path-noscheme = segment-nz-nc *( "/" segment )
path-empty    = 0<pchar>

segment       = *pchar
segment-nz    = 1*pchar
segment-nz-nc = 1*( unreserved / pct-encoded / sub-delims / "@" )
```

In [RFC3986.5](https://tools.ietf.org/html/rfc3986#section-5) defines the process of resolving a URI reference within a context that allows relative references.

The base URI of a reference can be established in one of four ways.
1. Base URI Embedded in Content
2. Base URI from the Encapsulating Entity
3. Base URI from the Retrieval URI
4. Default Base URI

We did not embed or encapsulate the base URI, so that would leave option **3** or **4**.

```
All requests to the service provider are made via HTTP methods as per
Section 4.3 of RFC7231 on a URL derived from the Base URL.
```

I interpreted this so that every endpoint is defined against the same base URL and thus clearly derivable. This would defining a base URL obsolete.

Are there cases where this cannot be derived that easily?

becoded replied:
You are right, we missed the partial-URI part....
I reverted the basePath changes.

di-wu commented on service_provider_config.go:
```diff
@@ -10,6 +12,9 @@ type ServiceProviderConfig struct {
 	DocumentationURI optional.String
+	// BasePathResolver is used to be able to handle requests with a common prefix for the exposed endpoints
+	// (Schema, ResourceTypes, ServiceProviderConfig & configured ResourceType endpoints.)
+	BasePathResolver func(r *http.Request) string
```
I would prefer not to include this in the library (see previous note).

becoded commented on .gitignore:
```diff
@@ -0,0 +1,56 @@
+*.msp
+
+# Windows shortcuts
+*.lnk
```
Np, will remove it again.

becoded commented on handlers.go:
```diff
@@ -165,13 +173,19 @@ func (s Server) resourcePatchHandler(w http.ResponseWriter, r *http.Request, id
+
+	if len(resource.Meta.Version) != 0 {
+		w.Header().Set("Etag", resource.Meta.Version)
+	}
```
👍

becoded commented:
- Did an update to use testify in the handler_test.
- Removed .gitignore file
- Change if-statement to make it more readable
- Added reasoning for the basePath change.

So ready for another review.

di-wu commented on resource_handler.go:
```diff
@@ -27,12 +28,21 @@ type ListRequestParams struct {
 type ResourceAttributes map[string]interface{}
 
+// Meta represents the metadata of a resource
+type Meta struct {
+	Created      *time.Time
+	LastModified *time.Time
+	Version      string
+}
```
One last thing, could you add comments to the exported fields of `Meta`?

becoded replied:
Done!

becoded commented:
I just fixed the merge conflicts.
